datasets/rwth_phoenix_weather_2014_t/rwth_phoenix_weather_2014_t.py

```diff
@@ -0,0 +1,112 @@
+  pages={7784--7793},
+  year={2018}
+}
+@article{koller2015continuous,
+  title={Continuous sign language recognition: 
+  Towards large vocabulary statistical recognition systems handling multiple signers},
+  author={Koller, Oscar and Forster, Jens and Ney, Hermann},
+  journal={Computer Vision and Image Understanding},
+  volume={141},
+  pages={108--125},
+  year={2015},
+  publisher={Elsevier}
+}
+"""
+
+_URL = "ftp://wasserstoff.informatik.rwth-aachen.de/pub/rwth-phoenix/2016/phoenix-2014-T.v3.tar.gz"
+
+_HOMEPAGE = "https://www-i6.informatik.rwth-aachen.de/~koller/RWTH-PHOENIX-2014-T/"
+
+
+class RWTHPhoenixWeather2014T(datasets.GeneratorBasedBuilder):
+    """RWTH-PHOENIX-Weather 2014 T: Parallel Corpus of Sign Language Video, Gloss and Translation"""
+
+    VERSION = datasets.Version("1.0.0")
+
+    def _info(self):
+        # TODO: Specifies the datasets.DatasetInfo object
+        return datasets.DatasetInfo(
+            description=_DESCRIPTION,
+            # This defines the different columns of the dataset and their types
+            features=datasets.Features({
+                "video": datasets.Value("string"),  # path to directory of images
+                "start": datasets.Value("int32"),  # starting frame of the annotation
+                "end": datasets.Value("int32"),  # ending frame of the annotation
+                "signer": datasets.Value("string"),  # signer ID
+                "gloss": datasets.Value("string"),  # German sign language gloss
+                "translation": datasets.Value("string")  # German translation
+            }),
+            homepage=_HOMEPAGE,
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager):
+        """Returns SplitGenerators."""
+
+        dl_dir = dl_manager.download_and_extract(_URL)
+        base_path = os.path.join(dl_dir, "PHOENIX-2014-T-release-v3", "PHOENIX-2014-T")
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                gen_kwargs={"base_path": base_path, "split": "train"},
+            ),
+            datasets.SplitGenerator(
+                name=datasets.Split.VALIDATION,
+                gen_kwargs={"base_path": base_path, "split": "dev"},
+            ),
+            datasets.SplitGenerator(
+                name=datasets.Split.TEST,
+                gen_kwargs={"base_path": base_path, "split": "test"},
+            ),
+        ]
+
+    def _generate_examples(self, base_path, split):
+        """ Yields examples. """
+
+        filepath = os.path.join(base_path, "annotations", "manual", "PHOENIX-2014-T." + split + ".corpus.csv")
+        images_path = os.path.join(base_path, "features", "fullFrame-210x260px", split)
+
+        with open(filepath) as f:
+            data = csv.DictReader(f, delimiter="|", quoting=csv.QUOTE_NONE)
+            for row in data:
+                yield row["name"], {
+                    "video": os.path.join(images_path, row["video"]),
```

Comment:
Do you think it could be worth adding the images directly in the samples instead of having the path ?

Reply:
That might be a good idea.
1. Could you point me to a dataset in this repo that you think does this well?
2. If the dataset contains images, would it still be able to load fast for people who only want to access `gloss` and `translation`, or would the images slow everything down drastically?

Reply:
We added support for images recently for the LXMERT dataset but it hasn't been added yet. Which means that this. dataset could be the first one with images :)

I've made images support compatible with memory mapping so you can load them very quickly.
Moreover since Apache Arrow has a NumPy integration it's even faster to load them if you format the dataset for numpy. It also makes loading as pytorch or tf tensors faster.

If your images are multi-channel high res and if you have several of them per samples, it will slow down accessing the other columns though. I haven't tested to have a dataset with a sequence of images per sample.

Could you tell me more about the images resolutions / nb of channels / nb of images per sample ?

Reply:
Pinging @sgugger here who know a lot about the tradeoffs here

Reply:
The images are `210x260`, `RGB` channels, with up to (maximum) 400 images per sample.
It would be equivalent in space to a `2184x10000` high res image.

I guess I'll give that a go unless @sgugger knows why I shouldn't

Reply:
Images are usually stored as encoded jpegs (or other format) to avoid the dataset being too large. I don't know how many images this dataset contains but keep in mind storing them in raw format could easily blow up disk storage. But if it fits and is not a problem for the size of the download, it would certainly speed up the decoding part.
 
Of course that also means you should avoid loading everything in RAM but I think the datasets library is already doing this?

Reply:
I couldn't find a feature that is specifically for images, in order to store it in a compressed format encoding (jpeg etc..)
Right now, I'm thinking on doing `Sequence(Array3D(shape=...))`
@lhoestq does this makes sense or is there a specific type for it?

Reply:
> Of course that also means you should avoid loading everything in RAM but I think the datasets library is already doing this?

Yes indeed, since the dataset is memory mapped.



> I couldn't find a feature that is specifically for images, in order to store it in a compressed format encoding (jpeg etc..)
> Right now, I'm thinking on doing `datasets.features.Sequence(datasets.features.Array3D(shape=(210, 260, 3), dtype="uint8"))`
> @lhoestq does this makes sense or is there a specific type for it?

That's the right feature type for your case.

Reply:
Thanks @lhoestq 

I now added this type, as well as am loading the images to numpy array in the correct shape, but this fails for some reason and I can't figure out why.
I am setting `video` as a list (sequential) of 3D numpy arrays (shape `260, 210, 3`, type `uint8`)

Mind taking a look? You can change the dataset path to the `dummy_data.zip` file, which includes 1 item, and would be very fast to load and process.

```
Traceback (most recent call last):
  File "/home/nlp/amit/forks/datasets/src/datasets/builder.py", line 434, in incomplete_dir
    yield tmp_dir
  File "/home/nlp/amit/forks/datasets/src/datasets/builder.py", line 476, in download_and_prepare
    dl_manager=dl_manager, verify_infos=verify_infos, **download_and_prepare_kwargs
  File "/home/nlp/amit/forks/datasets/src/datasets/builder.py", line 553, in _download_and_prepare
    self._prepare_split(split_generator, **prepare_split_kwargs)
  File "/home/nlp/amit/forks/datasets/src/datasets/builder.py", line 846, in _prepare_split
    num_examples, num_bytes = writer.finalize()
  File "/home/nlp/amit/forks/datasets/src/datasets/arrow_writer.py", line 299, in finalize
    self.write_on_file()
  File "/home/nlp/amit/forks/datasets/src/datasets/arrow_writer.py", line 232, in write_on_file
    pa_array = pa.array(typed_sequence)
  File "pyarrow/array.pxi", line 196, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 107, in pyarrow.lib._handle_arrow_array_protocol
  File "/home/nlp/amit/forks/datasets/src/datasets/arrow_writer.py", line 99, in __arrow_array__
    out = pa.array(self.data, type=type)
  File "pyarrow/array.pxi", line 269, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 38, in pyarrow.lib._sequence_to_array
  File "pyarrow/error.pxi", line 105, in pyarrow.lib.check_status
pyarrow.lib.ArrowNotImplementedError: MakeBuilder: cannot construct builder for type extension<arrow.py_extension_type>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/nlp/amit/forks/datasets/src/load_test.py", line 6, in <module>
    '/home/nlp/amit/forks/datasets/datasets/rwth_phoenix_weather_2014_t/rwth_phoenix_weather_2014_t.py')
  File "/home/nlp/amit/forks/datasets/src/datasets/load.py", line 611, in load_dataset
    ignore_verifications=ignore_verifications,
  File "/home/nlp/amit/forks/datasets/src/datasets/builder.py", line 483, in download_and_prepare
    self._save_info()
  File "/home/nlp/amit/anaconda2/envs/meta-scholar/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/nlp/amit/forks/datasets/src/datasets/builder.py", line 440, in incomplete_dir
    shutil.rmtree(tmp_dir)
  File "/home/nlp/amit/anaconda2/envs/meta-scholar/lib/python3.7/shutil.py", line 498, in rmtree
    onerror(os.rmdir, path, sys.exc_info())
  File "/home/nlp/amit/anaconda2/envs/meta-scholar/lib/python3.7/shutil.py", line 496, in rmtree
    os.rmdir(path)
OSError: [Errno 39] Directory not empty: '/home/nlp/amit/.cache/huggingface/datasets/rwth_phoenix_weather2014_t/default/1.0.0/f94e5a4ba5905587046b092078275d655495cb484f281c43d3a53f85ac366743.incomplete'
```